src/fields/surface_depth.cc

```diff
@@ -0,0 +1,158 @@
+    		    			<< EI_Message("Dimension of surface region " + surface_region + " must be 2!") );
+    }
+
+    // make element boxes
+    unsigned int i=0;
+    unsigned int i_node;
+    arma::vec3 project_node("0 0 0");
+    for( ElementFullIter it = mesh->bc_elements.begin(); it != mesh->bc_elements.end(); ++it) {
+        if (it->region().is_in_region_set(region_set)) {
+        	ASSERT_EQ(it->n_nodes(), 3);
+
+        	arma::mat coords(3,3);
+        	coords.col(0) = it->node[0]->point();
+        	arma::vec projection = m_ * it->node[0]->point();
+        	project_node(0) = projection(0); project_node(1) = projection(1);
+            BoundingBox bb(project_node);
+            for(i_node=1; i_node<it->n_nodes(); i_node++) {
+            	coords.col(i_node) = it->node[i_node]->point();
+                arma::vec project_coords = m_ * it->node[i_node]->point();
+                project_node(0) = project_coords(0); project_node(1) = project_coords(1);
+                bb.expand(project_node);
+            }
+            boxes.push_back(bb);
+            nodes_coords_.push_back(coords);
+        }
+        i++;
+    }
+
+    if ( boxes.size() == 0) {
+    	THROW( ExcSurfaceProjection()
+    			<< EI_Message("Region " + surface_region + " contains no boundary element! Probably bulk region was set.") );
+    }
+
+    bih_tree_.add_boxes( boxes );
+    bih_tree_.construct();
+}
+
+double distance_from_abscissa(arma::vec3 a, arma::vec3 b, arma::vec3 &p)
+{
+	arma::vec3 u = b-a;
+	double d = - arma::dot( u, p );
+	double t = - ( a(0)*u(0) + a(1)*u(1) + a(2)*u(2) + d) / ( u(0)*u(0) + u(1)*u(1) + u(2)*u(2) );
+	if (t<0.0) t=0.0;
+	else if (t>1.0) t=1.0;
+
+	arma::vec3 insec;
+	for (unsigned int i=0;i<3;++i) insec(i) = a(i) + u(i) * t;
+	return arma::norm( (insec-p), 2 );
+}
+
+double SurfaceDepth::compute_distance(arma::vec3 point)
+{
+	double distance = std::numeric_limits<double>::max();
+	double snap_dist = std::numeric_limits<double>::max();
+	arma::vec3 project_point;
+	arma::vec3 proj_to_surface_plane;
+	project_point.subvec(0,1) = m_ * point;
+	project_point(2) = 0;
+
+	std::vector<unsigned int> searched_elements;
+	bih_tree_.find_point(project_point, searched_elements);
+	for (std::vector<unsigned int>::iterator it = searched_elements.begin(); it!=searched_elements.end(); it++) {
+		auto coords = nodes_coords_[(*it)];
+		arma::vec3 va = coords.col(1) - coords.col(0);
+		arma::vec3 vb = coords.col(2) - coords.col(0);
+
+		arma::mat a_mat(3,3);
+		a_mat.col(0) = va;
+		a_mat.col(1) = vb;
+		a_mat.col(2) = surface_norm_vec_;
+		arma::vec3 b_vec = point - coords.col(0);
+		arma::vec3 x;
+		arma::solve(x, a_mat, b_vec);
+		if ( (x(0)>=0) && (x(1)>=0) && (x(0)+x(1)<=1) ) { // point is in triangle
+			double new_distance = -x(2);
+			if ( (new_distance>=0) && (new_distance<distance) || (snap_dist>0.0) ) distance = new_distance;
+			snap_dist = 0.0;
+		} else if (snap_dist > 0.0) { // check snap distance of point from triangle
```

Comment:
Perform clipping only if no element is found. Second loop over searched_elements